### DISCUSSION TITLE
Assets field - make _getUploadedFiles publicly available / protected instead of private

### DISCUSSION THREAD

**Discussion by Anubarak** [craftcms/cms#11123] [Ideas] [answered]
I very often have the requirement to upload files via Dropzone.js or other JS libraries in forms before the final form submission. So users can upload files to the server **before** the related element is uploaded. There are only a few ways to achieve this.

- upload the file normally, create an asset -> return the id
  -   disadvantages: 
      - Assets are always saved on the server, even if there is no element related to them if the user decides to just leave the site
      - Craft does not handle proper sub-folder logic like the field would -> the developer needs to create subfolders - otherwise all are in the very same - this is even more frustrating since the logic is already there... just private
- use a normal `<input type="file">` and abuse Dropzone (or similar libraries) in a way it can insert the file into an `<input type="file">` or into a hidden input with the base64 encoded content in the end.
    - disadvantages
        - well obviously not using libraries kinda eliminates their purpose + uploading the file before will make the request much faster

My wish would be to be able to upload files normally however I like, move them to a temp folder of my choice (in my custom logic) and then just hook into `craft\fields\Assets::_getUploadedFiles` in order to include custom data like you do at the end of the function
```PHP
$files = UploadedFile::getInstancesByName($paramName);

foreach ($files as $file) {
    $uploadedFiles[] = [
        'filename' => $file->name,
        'location' => $file->tempName,
        'type' => 'custom', // use a custom type here
    ];
}
```

it would even be perfect to extend your `Assets` field and add an event or a custom check into this function but since it's private I can't call the parents method... unless I move all the logic into my own class.

Eventually the `afterElementsSave` must be changed as well to add
```PHP
if ($file['type'] === 'custom') {
    copy($file['location'], $tempPath);
}
```

Of course in the end of the day this is still kinda an ugly way but at least this one achieves my goal without heavy changes on your end... A better way would be to include more events or move all private functions into services so I could do
```PHP
$entry->on(Element::EVENT_AFTER_PROPAGATE, static function(ModelEvent $event){
    $entry = $event->sender;
    $field = $entry->getFieldLayout()->getFieldByHandle('image');
    $targetFolderId = $field->determineUploadFolderId($entry);
    
    // save the image by myself.. I just want that folder...
});
\Craft::$app->getElements()->saveElement($entry);
```

**Comment by brandonkelly** [comment, marked as answer]
Just added a new `EVENT_LOCATE_UPLOADED_FILES` event to Assets fields, which you can tap into to add additional files to the mix, without needing to extend `craft\fields\Assets`.

```php
use craft\events\LocateUploadedFilesEvent;
use craft\fields\Assets;
use yii\base\Event;

Event::on(
    Assets::class,
    Assets::EVENT_LOCATE_UPLOADED_FILES,
    function(LocateUploadedFilesEvent $event) {
        /** @var Assets $field */
        $field = $event->sender;
        
        $event->files[] = [
            'type' => 'file',
            'filename' => 'foo.jpg',
            'path' => '/path/to/temp/foo.jpg',
        ];
    }
);
```

It’ll be in the next v3 and v4 releases.

To get the change early, you can change your `craftcms/cms` requirement in `composer.json` to one of these:

- `dev-release/3.7.42 as 3.7.42`
- `dev-release/4.0.2 as 4.0.2`

**Reply by Anubarak** [reply]
Awesome, thank you very much for this 👍

**Reply by brandonkelly** [reply]
Craft 3.7.42 and 4.0.2 are out now with that change.
